### ISSUE TITLE
minifab fails to run on Mac OS X

### ISSUE THREAD

**Issue by atoulme** [hyperledger-labs/minifabric#141]
There seems to be a problem with installing chaincode on Mac OS with a simple `minifab up` command:

Steps to reproduce:
```
mkdir -p vars
curl -o minifab -sL https://tinyurl.com/yxa2q6yr && chmod +x minifab
./minifab up --fabric-release 2.2
```

```
Using default spec file
Minifab Execution Context:
    FABRIC_RELEASE=2.2
    CHANNEL_NAME=mychannel
    PEER_DATABASE_TYPE=golevel
    CHAINCODE_LANGUAGE=go
    CHAINCODE_NAME=simple
    CHAINCODE_VERSION=1.0
    CHAINCODE_INIT_REQUIRED=true
    CHAINCODE_PARAMETERS="init","a","200","b","300"
    CHAINCODE_PRIVATE=false
    CHAINCODE_POLICY=
    TRANSIENT_DATA=
    BLOCK_NUMBER=newest
    EXPOSE_ENDPOINTS=false
    CURRENT_ORG=org0.example.com
    HOST_ADDRESSES=10.132.125.161,192.168.88.19
    WORKING_DIRECTORY: /Users/atoulme/w/fabric-logger/examples/minifab
....
# Preparing for the following operations: *********************
  verify options, download images, generate certificates, start network, network status, channel create, channel join, anchor update, profile generation, cc install, cc approve, cc commit, cc initialize, discover
.................
# Running operation: ******************************************
  verify options
.
# Running operation: ******************************************
  download images
..............
# Running operation: ******************************************
  generate certificates
............................................................................................................................................
# Running operation: ******************************************
  start network
.......................
# Running operation: ******************************************
  network status
......
# Docker node status ******************************************
  119542aca2_cli : Up 6 seconds
  ca1.org1.example.com : Up 7 seconds
  ca1.org0.example.com : Up 7 seconds
  orderer3.example.com : Up 8 seconds
  orderer2.example.com : Up 9 seconds
  orderer1.example.com : Up 10 seconds
  peer2.org1.example.com : Up 10 seconds
  peer1.org1.example.com : Up 11 seconds
  peer2.org0.example.com : Up 12 seconds
  peer1.org0.example.com : Up 13 seconds

# Fabric network peer and orderer node health status **********
  peer1.org0.example.com 
  peer2.org0.example.com 
  peer1.org1.example.com 
  peer2.org1.example.com 
  orderer1.example.com "OK"
  orderer2.example.com "OK"
  orderer3.example.com "OK"

# Running operation: ******************************************
  channel create
......
# Running operation: ******************************************
  channel join
...........................
# Running operation: ******************************************
  anchor update
............
# Running operation: ******************************************
  profile generation
.................................
# Running operation: ******************************************
  cc install
.......
# Run the chaincode install script on cli container ***********
  non-zero return code
  go: downloading github.com/hyperledger/fabric v1.4.1
  go: downloading github.com/golang/protobuf v1.3.1
  go: finding module for package github.com/op/go-logging
  go: finding module for package github.com/pkg/errors
  go: finding module for package golang.org/x/net/context
  go: finding module for package google.golang.org/grpc
  go: finding module for package github.com/grpc-ecosystem/go-grpc-middleware
  go: finding module for package go.uber.org/zap/zapcore
  go: finding module for package google.golang.org/grpc/credentials
  go: finding module for package go.uber.org/zap
  go: downloading google.golang.org/grpc v1.34.0
  go: downloading github.com/grpc-ecosystem/go-grpc-middleware v1.2.2
  go: downloading go.uber.org/zap v1.16.0
  go: downloading github.com/pkg/errors v0.9.1
  go: finding module for package google.golang.org/grpc/grpclog
  go: finding module for package google.golang.org/grpc/peer
  go: downloading github.com/op/go-logging v0.0.0-20160315200505-970db520ece7
  go: downloading golang.org/x/net v0.0.0-20201224014010-6772e930b67b
  go: finding module for package github.com/spf13/viper
  go: finding module for package google.golang.org/grpc/stats
  go: finding module for package golang.org/x/crypto/sha3
  go: downloading github.com/spf13/viper v1.7.1
  go: finding module for package github.com/sykesm/zap-logfmt
  go: finding module for package github.com/miekg/pkcs11
  go: downloading github.com/sykesm/zap-logfmt v0.0.4
  go: finding module for package go.uber.org/zap/zapgrpc
  go: downloading golang.org/x/crypto v0.0.0-20201221181555-eec23a3978ad
  go: downloading github.com/miekg/pkcs11 v1.0.3
  go: finding module for package google.golang.org/grpc/keepalive
  go: finding module for package github.com/fsouza/go-dockerclient
  go: finding module for package go.uber.org/zap/buffer
  go: finding module for package gopkg.in/yaml.v2
  go: finding module for package github.com/hyperledger/fabric-amcl/amcl
  go: finding module for package github.com/hyperledger/fabric-amcl/amcl/FP256BN
  go: downloading gopkg.in/yaml.v2 v2.4.0
  go: downloading github.com/fsouza/go-dockerclient v1.7.0
  go: downloading github.com/hyperledger/fabric-amcl v0.0.0-20200424173818-327c9e2cf77a
  go: found github.com/op/go-logging in github.com/op/go-logging v0.0.0-20160315200505-970db520ece7
  go: found github.com/pkg/errors in github.com/pkg/errors v0.9.1
  go: found github.com/spf13/viper in github.com/spf13/viper v1.7.1
  go: found google.golang.org/grpc in google.golang.org/grpc v1.34.0
  go: found golang.org/x/net/context in golang.org/x/net v0.0.0-20201224014010-6772e930b67b
  go: found github.com/grpc-ecosystem/go-grpc-middleware in github.com/grpc-ecosystem/go-grpc-middleware v1.2.2
  go: found github.com/miekg/pkcs11 in github.com/miekg/pkcs11 v1.0.3
  go: found go.uber.org/zap/zapcore in go.uber.org/zap v1.16.0
  go: found golang.org/x/crypto/sha3 in golang.org/x/crypto v0.0.0-20201221181555-eec23a3978ad
  go: found github.com/sykesm/zap-logfmt in github.com/sykesm/zap-logfmt v0.0.4
  go: found gopkg.in/yaml.v2 in gopkg.in/yaml.v2 v2.4.0
  go: found github.com/fsouza/go-dockerclient in github.com/fsouza/go-dockerclient v1.7.0
  go: found github.com/hyperledger/fabric-amcl/amcl in github.com/hyperledger/fabric-amcl v0.0.0-20200424173818-327c9e2cf77a
  go: downloading github.com/golang/protobuf v1.4.2
  go: downloading golang.org/x/text v0.3.3
  go: downloading google.golang.org/genproto v0.0.0-20200526211855-cb27e3aa2013
  go: downloading go.uber.org/atomic v1.6.0
  go: downloading google.golang.org/protobuf v1.25.0
  go: downloading go.uber.org/multierr v1.5.0
  go: downloading github.com/spf13/afero v1.1.2
  go: downloading github.com/hashicorp/hcl v1.0.0
  go: downloading github.com/spf13/jwalterweatherman v1.0.0
  go: downloading github.com/fsnotify/fsnotify v1.4.7
  go: downloading golang.org/x/sys v0.0.0-20201119102817-f84b799fce68
  go: downloading github.com/spf13/cast v1.3.0
  go: downloading gopkg.in/ini.v1 v1.51.0
  go: downloading github.com/docker/go-units v0.4.0
  go: downloading github.com/subosito/gotenv v1.2.0
  go: downloading github.com/docker/docker v20.10.0-beta1.0.20201113105859-b6bfff2a628f+incompatible
  go: downloading github.com/Microsoft/go-winio v0.4.16
  go: downloading github.com/mitchellh/mapstructure v1.1.2
  go: downloading github.com/pelletier/go-toml v1.2.0
  go: downloading github.com/magiconair/properties v1.8.1
  go: downloading github.com/spf13/pflag v1.0.3
  go: downloading github.com/sirupsen/logrus v1.6.0
  go: downloading github.com/moby/term v0.0.0-20201110203204-bea5bbe245bf
  go: downloading github.com/opencontainers/image-spec v1.0.1
  go: downloading github.com/morikuni/aec v1.0.0
  go: downloading github.com/gogo/protobuf v1.3.1
  go: downloading github.com/docker/go-connections v0.4.0
  go: downloading github.com/containerd/continuity v0.0.0-20200928162600-f2cc35102c2a
  go: downloading github.com/opencontainers/runc v0.1.1
  go: downloading github.com/Microsoft/hcsshim v0.8.10
  go: downloading github.com/moby/sys/mount v0.2.0
  go: downloading github.com/containerd/containerd v1.4.1
  go: downloading github.com/Azure/go-ansiterm v0.0.0-20170929234023-d6e3b3328b78
  go: downloading github.com/konsorten/go-windows-terminal-sequences v1.0.3
  go: downloading golang.org/x/sync v0.0.0-20190423024810-112230192c58
  go: downloading github.com/moby/sys/mountinfo v0.4.0
  go: downloading github.com/opencontainers/go-digest v1.0.0
  go: downloading go.opencensus.io v0.22.0
  go: downloading github.com/containerd/cgroups v0.0.0-20200531161412-0dbf7f05ba59
  go: downloading github.com/hashicorp/golang-lru v0.5.1
  Error: chaincode install failed with status: 500 - failed to invoke backing implementation of 'InstallChaincode': could not build chaincode: docker build failed: docker image inspection failed: Get "http://unix.sock/images/dev-peer1.org0.example.com-simple_1.0-27ce0a92733f636acbc6e04da021fe77279a6018fb960e970bae00c1e0617a8c-0c1f096218a1f8b48ae2292827925e7d83c40fc5b66d22191e465100bbf2fe09/json": dial unix /host/var/run/docker.sock: connect: no such file or directory

# STATS *******************************************************
minifab: ok=338 failed=1        

real    2m35.590s
user    1m30.681s
sys     0m25.901s
```

**Comment by litong01** [comment]
@atoulme were you able to do the same in Ubuntu or Windows env.? What version of docker desktop do you have with OS X?

**Comment by litong01** [comment]
@atoulme I've tested it on my mac, it worked fine. i am thinking that you may have network issues. Looking at your screen capture, your HOST_ADDRESSES look like this

`HOST_ADDRESSES=10.132.125.161,192.168.88.19`

These are `minifabric` auto detected IP addresses. it is possible that 10.132.125.161 may be restricted (or firewalled), you probably can try to explicitly specify your IP in spec.yaml file.

basically just copy and paste this file `https://github.com/hyperledger-labs/minifabric/blob/master/spec.yaml` to your work directory but uncomment this line 

` # endpoint_address: 1.2.3.4`

and use your own IP address such as `192.168.88.19`, this way, minifabric will use that IP address and hoping that ip was not firewalled.

**Comment by atoulme** [comment]
I do not have a Linux or Windows box available for this exercise. I am trying to create a working Fabric sample for developers on Mac OS X. I run Docker `v20.10.0`.

I followed the instructions you referenced and reproduced the same error.

**Comment by litong01** [comment]
@atoulme I think you have to disable the Use gRPC FUSE for file sharing. This is known issue. Go into docker dashboard, then Experimental Features, make sure you have Use gRPC FUSE for file sharing unchecked.

**Comment by litong01** [comment]
By known issue, I mean, it is an issue with docker in general, not minifabric.

**Comment by atoulme** [comment]
Indeed, this was the problem. Thank you. I will reference this issue in our notes.

**Comment by litong01** [comment]
In that case, please close this issue.

**Comment by atoulme** [comment]
That doesn't seem adequate to me ; you provided a workaround, and it fixes the issue at this time. However, this workaround may expire going forward and it seems likely a better fix is required here. At the very least, documenting this problem in the documentation seems indicated. You may of course differ in opinion, and decide to close this issue of your own accord.

**Comment by litong01** [comment]
@atoulme there is nothing I can do with this issue. This is not a workaround, you are using an experimental feature which causes problems specifically for OS X and most likely won't be a targeting env for any production. 

If you want to contribute a PR for what you suggested, you are welcome to do so, please put up a fix and I will review it.

The point is that we cannot leave these issues open without taking care of it.

**Comment by awattez** [comment]
unchecked gRPC FUSE fixe same problem but i have another one after that:
```bash
~/mywork|⇒  ./minifab up
Using default spec file
Minifab Execution Context:
    FABRIC_RELEASE=2.3.0
    CHANNEL_NAME=mychannel
    PEER_DATABASE_TYPE=golevel
    CHAINCODE_LANGUAGE=go
    CHAINCODE_NAME=simple
    CHAINCODE_VERSION=1.0
    CHAINCODE_INIT_REQUIRED=true
    CHAINCODE_PARAMETERS="init","a","200","b","300"
    CHAINCODE_PRIVATE=false
    CHAINCODE_POLICY=
    TRANSIENT_DATA=
    BLOCK_NUMBER=newest
    EXPOSE_ENDPOINTS=false
    CURRENT_ORG=org0.example.com
    HOST_ADDRESSES=192.168.1.39
    WORKING_DIRECTORY: /Users/awattez/mywork
....
# Preparing for the following operations: *********************
  verify options, download images, generate certificates, start network, network status, channel create, channel join, anchor update, profile generation, cc install, cc approve, cc commit, cc initialize, discover
.......
# Running operation: ******************************************
  verify options
.
# Running operation: ******************************************
  download images
............
# Running operation: ******************************************
  generate certificates
..
# Running operation: ******************************************
  start network
...................
# Running operation: ******************************************
  network status
...
# Current Minifabric image ID and created at date time ********
  998211d22979   2020-12-18 05:34:27 +0000 UTC
.....
# Docker node status ******************************************
  2a8155d0b3_cli : Up 7 seconds
  ca1.org1.example.com : Up 8 seconds
  ca1.org0.example.com : Up 9 seconds
  orderer3.example.com : Up 10 seconds
  orderer2.example.com : Up 11 seconds
  orderer1.example.com : Up 12 seconds
  peer2.org1.example.com : Up 13 seconds
  peer1.org1.example.com : Up 14 seconds
  peer2.org0.example.com : Up 15 seconds
  peer1.org0.example.com : Up 15 seconds

# Fabric network peer and orderer node health status **********
  peer1.org0.example.com "OK"
  peer2.org0.example.com "OK"
  peer1.org1.example.com "OK"
  peer2.org1.example.com "OK"
  orderer1.example.com "OK"
  orderer2.example.com "OK"
  orderer3.example.com "OK"
  Network Status: 100%

# Running operation: ******************************************
  channel create
.....
# Run the channel creation script on cli container ************
  non-zero return code
  2021-01-21 22:25:44.585 UTC [common.tools.configtxgen] main -> INFO 001 Loading configuration
  2021-01-21 22:25:44.618 UTC [common.tools.configtxgen.localconfig] Load -> INFO 002 Loaded configuration: /vars/configtx.yaml
  2021-01-21 22:25:44.618 UTC [common.tools.configtxgen] doOutputChannelCreateTx -> INFO 003 Generating new channel configtx
  2021-01-21 22:25:44.815 UTC [common.tools.configtxgen] doOutputChannelCreateTx -> INFO 004 Writing new channel tx
  2021-01-21 22:25:44.999 UTC [channelCmd] InitCmdFactory -> INFO 001 Endorser and orderer connections initialized
  Error: got unexpected status: BAD_REQUEST -- error applying config update to existing channel 'mychannel': error authorizing update: error validating ReadSet: proposed update requires that key [Group]  /Channel/Application be at version 0, but it is currently at version 1

# STATS *******************************************************
minifab: ok=85	failed=1

real	0m37.980s
user	0m22.110s
sys	0m6.602s
```

**Comment by litong01** [comment]
I suspect your env is not clean. Try minifab cleanup then rerun.

**Comment by awattez** [comment]
`minifab cleanup` then `minifab up` works.
Thanks

**Comment by litong01** [comment]
Please close this issue.

**Comment by ryjones** [comment]
@atoulme I was researching this issue - I [found this](https://github.com/docker/for-mac/issues/5124), which said it's fixed in Desktop 3.0.4. I installed the current release, Desktop 3.1.0, which installs 20.10.2:

`% docker --version`
`Docker version 20.10.2, build 2291f61`

`minifab up` works as expected, even with experimental gRPC-FUSE feature enabled. I'm running `macOS 11.1 (20C69)` on an Intel mac.

**Comment by litong01** [comment]
@ryjones I've upgraded to the version you mentioned and tried the simple command, it still fails during the chaincode install. Something is still not right. Disabling the feature, everything worked fine.
